src/secret_storage.py

```
@@ -0,0 +1,90 @@
+    # please check the docs:
+    # https://docs.python.org/3/library/secrets.html#how-many-bytes-should-tokens-use
+    username = token_hex(16)
+    if JujuVersion.from_environ().has_secrets and not peer_relation.data[charm.app].get(SECRET_ID):
+        # we can create secrets and the one that we need was not created yet
+        logger.debug("Adding secret")
+        admin_user = actions.register_user(container, username, True)
+        secret = charm.app.add_secret({SECRET_KEY: admin_user.access_token})
+        peer_relation.data[charm.app].update({SECRET_ID: secret.id})
+        return
+
+    if not JujuVersion.from_environ().has_secrets and not peer_relation.data[charm.app].get(
+        SECRET_KEY
+    ):
+        # we can't create secrets and peer data is empty
+        logger.debug("Updating peer relation data")
+        admin_user = actions.register_user(container, username, True)
+        peer_relation.data[charm.app].update({SECRET_KEY: admin_user.access_token})
+
+
+def get_admin_access_token(charm: ops.CharmBase) -> typing.Optional[str]:
+    """Get admin access token.
+
+    Args:
+        charm: The charm object.
+
+    Returns:
+        admin access token.
+    """
+    container = charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+    if not container.can_connect():
+        logger.warning("Failed to get secret storage: waiting for pebble")
+        return None
+    peer_relation = charm.model.get_relation(PEER_RELATION_NAME)
+    if not peer_relation:
+        # there is no peer relation so nothing to be done
+        logger.warning("Failed to get secret storage: waiting for peer relation")
+        return None
+    _update_peer_data(charm, container)
+    if JujuVersion.from_environ().has_secrets:
+        secret_id = peer_relation.data[charm.app].get(SECRET_ID)
+        if secret_id:
+            secret = charm.model.get_secret(id=secret_id)
+            secret_value = secret.get_content().get(SECRET_KEY)
```

Should probably thrown an exception is the secret_id is not found, and also do some cleanup in the peer databag

---

Exception is being raised now.
What do you mean by cleanup? If secret id is not found, what should be done?

---

If we have a secret_id in the peer data relation that doesn't exists as a secret, so we should remove that secret_id from the peer relation data

---

Oh now I see...Well, in that case, shouldn't we let the charm raise an error? because if there is an ID, something happened outside of the charm to delete the expected secret.

---

I just checked and if the secret is not found a SecretNotFound exception is raised. My worry is that since the secret_id will stay in the peer relation data and I don't know what the user can do to cleanup these data (unless undeploy / redeploy).
I think that's better to catch the error, cleanup the peer relation data and reraise the error

---

Nice, changed.